core/commands/cid.go

```diff
@@ -0,0 +1,336 @@
+	},
+	Options: []cmdkit.Option{
+		cmdkit.StringOption("f", "Printf style format string.").WithDefault("%s"),
+		cmdkit.StringOption("v", "CID version to convert to."),
+		cmdkit.StringOption("b", "Multibase to display CID in."),
+	},
+	Run: func(req *cmds.Request, resp cmds.ResponseEmitter, env cmds.Environment) {
+		fmtStr, _ := req.Options["f"].(string)
+		verStr, _ := req.Options["v"].(string)
+		baseStr, _ := req.Options["b"].(string)
+
+		opts := cidFormatOpts{}
+
+		if strings.IndexByte(fmtStr, '%') == -1 {
+			resp.SetError(fmt.Errorf("invalid format string: %s", fmtStr), cmdkit.ErrNormal)
+			return
+		}
+		opts.fmtStr = fmtStr
+
+		switch verStr {
+		case "":
+			// noop
+		case "0":
+			opts.verConv = toCidV0
+		case "1":
+			opts.verConv = toCidV1
+		default:
+			resp.SetError(fmt.Errorf("invalid cid version: %s\n", verStr), cmdkit.ErrNormal)
+			return
+		}
+
+		if baseStr != "" {
+			encoder, err := mbase.EncoderByName(baseStr)
+			if err != nil {
+				resp.SetError(err, cmdkit.ErrNormal)
+				return
+			}
+			opts.newBase = encoder.Encoding()
+		} else {
+			opts.newBase = mbase.Encoding(-1)
+		}
+
+		emitCids(req, resp, opts)
+	},
+	PostRun: cmds.PostRunMap{
+		cmds.CLI: streamRes(func(v interface{}, out io.Writer) nonFatalError {
+			r := v.(*CidFormatRes)
+			if r.ErrorMsg != "" {
+				return nonFatalError(fmt.Sprintf("%s: %s", r.CidStr, r.ErrorMsg))
+			}
+			fmt.Fprintf(out, "%s\n", r.Formatted)
+			return ""
+		}),
+	},
+	Type: CidFormatRes{},
+}
+
+type CidFormatRes struct {
+	CidStr    string // Original Cid String passed in
+	Formatted string // Formated Result
+	ErrorMsg  string // Error
+}
+
+var base32Cmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline: "Convert CIDs to Base32 CID version 1.",
+	},
+	Arguments: []cmdkit.Argument{
+		cmdkit.StringArg("cid", true, true, "Cids to convert.").EnableStdin(),
+	},
+	Run: func(req *cmds.Request, resp cmds.ResponseEmitter, env cmds.Environment) {
+		opts := cidFormatOpts{
+			fmtStr:  "%s",
+			newBase: mbase.Encoding(mbase.Base32),
+			verConv: toCidV1,
+		}
+		emitCids(req, resp, opts)
+	},
+	PostRun: cidFmtCmd.PostRun,
+	Type:    cidFmtCmd.Type,
+}
+
+type cidFormatOpts struct {
+	fmtStr  string
+	newBase mbase.Encoding
+	verConv func(cid *cid.Cid) (*cid.Cid, error)
+}
+
+type argumentIterator struct {
+	args []string
+	body cmds.StdinArguments
+}
+
+func (i *argumentIterator) next() (string, bool) {
+	if len(i.args) > 0 {
+		arg := i.args[0]
+		i.args = i.args[1:]
+		return arg, true
+	}
+	if i.body == nil || !i.body.Scan() {
+		return "", false
+	}
+	return strings.TrimSpace(i.body.Argument()), true
+}
+
+func (i *argumentIterator) err() error {
+	if i.body == nil {
+		return nil
+	}
+	return i.body.Err()
+}
+
+func emitCids(req *cmds.Request, resp cmds.ResponseEmitter, opts cidFormatOpts) {
+	itr := argumentIterator{req.Arguments, req.BodyArgs()}
+	for {
+		cidStr, ok := itr.next()
+		if !ok {
+			break
+		}
+		emit := func(fmtd string, err error) {
```

Comment:
Let's lift this out of the loop (I don't trust go).

Reply:
It's in the loop as it is a closure.  I do it that way to avoid having to pass is the cidStr variable.

Reply:
I just eliminated the need for the emit function.